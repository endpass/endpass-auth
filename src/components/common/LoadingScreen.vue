<template>
  <div v-if="isLoading">
    <div class="loading-screen">
      <div class="loading-screen__spinner">
        <Spinner />
      </div>
    </div>
  </div>
  <div v-else>
    <slot />
  </div>
</template>

<script>
import Spinner from '@/components/common/Spinner.vue';

export default {
  name: 'LoadingScreen',

  props: {
    isLoading: {
      type: Boolean,
      default: true,
    },
  },

  components: {
    Spinner,
  },
};
</script>

<style lang="postcss">
.loading-screen {
  padding: 30px 15px 15px;
  position: absolute;
  display: flex;
  align-items: center;
  justify-content: center;
  top: 0;
  left: 0;
  z-index: 2;
  width: 100%;
  height: 100%;
}

.loading-screen__spinner {
  flex: 0 0 auto;
}
</style>
