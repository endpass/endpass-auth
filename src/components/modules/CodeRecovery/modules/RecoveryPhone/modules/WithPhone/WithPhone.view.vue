<template>
  <form
    data-test="recover-otp"
    @submit.prevent="onSubmit"
  >
    <v-title>
      {{ $t('components.recoverOtpSms.title') }}
    </v-title>

    <form-item>
      <v-input
        v-model="code"
        v-validate="'required|digits:6'"
        autocomplete="off"
        data-vv-as="code"
        data-vv-name="code"
        :error="errors.first('code')"
        name="code"
        :placeholder="$t('components.appCode.enterReceivedCode')"
        data-test="code-input"
      />
    </form-item>
    <form-item class="v-mb-24">
      <form-controls>
        <v-button
          :disabled="isLoading"
          skin="quaternary"
          data-test="cancel-button"
          type="button"
          @click="onCancel"
        >
          {{ $t('global.cancel') }}
        </v-button>
        <v-button
          :disabled="!isFormValid || isLoading"
          :is-loading="isLoading"
          type="submit"
          data-test="submit-button"
        >
          {{ $t('global.confirm') }}
        </v-button>
      </form-controls>
    </form-item>
    <form-row class="v-fs-14 v-text-center">
      <v-link
        :disabled="isLoading"
        role="button"
        data-test="recovery-link"
        @click.prevent="onSendCode"
      >
        {{ $t('components.recoverOtpSms.sendTitle') }}
      </v-link>
    </form-row>
  </form>
</template>

<script>
import VButton from '@endpass/ui/kit/VButton';
import VInput from '@endpass/ui/kit/VInput';
import VLink from '@endpass/ui/kit/VLink';
import FormItem from '@/components/common/FormItem';
import FormRow from '@/components/common/FormRow';
import FormControls from '@/components/common/FormControls';
import formMixin from '@/mixins/form';
import VTitle from '@/components/common/VTitle';

export default {
  name: 'WithPhone',

  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },

    error: {
      type: String,
      default: '',
    },
  },

  data: () => ({
    code: '',
  }),

  watch: {
    error: {
      handler(msg) {
        this.$validator.errors.removeById('sendCodeId');

        if (!msg) return;

        this.$validator.errors.add({
          id: 'sendCodeId',
          field: 'code',
          msg,
        });
      },
      immediate: true,
    },
  },

  methods: {
    onSendCode() {
      this.$emit('send-code');
    },

    onSubmit() {
      this.$emit('submit', {
        code: this.code,
      });
    },

    onCancel() {
      this.$emit('cancel');
    },
  },

  mixins: [formMixin],

  components: {
    VTitle,
    VLink,
    VButton,
    VInput,
    FormItem,
    FormRow,
    FormControls,
  },
};
</script>
