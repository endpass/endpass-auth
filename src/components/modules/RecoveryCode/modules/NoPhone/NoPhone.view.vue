<template>
  <form
    data-test="phone-not-exist"
    @submit.prevent="onCancel"
  >
    <v-title>
      {{ $t('components.recoverOtpSms.titlePhoneNotExist') }}
    </v-title>

    <form-item class="v-mb-24">
      <v-button
        :is-loading="isLoading"
        type="submit"
        data-test="submit-button"
      >
        {{ $t('global.cancel') }}
      </v-button>
    </form-item>
  </form>
</template>

<script>
import VButton from '@endpass/ui/kit/VButton';
import FormItem from '@/components/common/FormItem';
import formMixin from '@/mixins/form';
import VTitle from '@/components/common/VTitle';

export default {
  name: 'NoPhone',

  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },
  },

  methods: {
    onCancel() {
      this.$emit('cancel');
    },
  },

  mixins: [formMixin],

  components: {
    VTitle,
    VButton,
    FormItem,
  },
};
</script>
