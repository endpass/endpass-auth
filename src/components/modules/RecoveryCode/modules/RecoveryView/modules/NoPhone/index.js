export { default } from './NoPhone.view';
