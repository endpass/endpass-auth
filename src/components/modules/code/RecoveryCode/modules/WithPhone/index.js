export { default } from './WithPhone.view';
