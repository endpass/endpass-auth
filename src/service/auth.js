// @ts-check
import get from 'lodash/get';
import request from '@/class/singleton/request';
import { AUTH_STATUS_CODE } from '@/constants';

const identityBaseUrl = ENV.VUE_APP_IDENTITY_API_URL;

const TIMEOUT_DEFAULT = 1500;

/**
 * @type {object}
 */
const CODE_TO_STATUS = {
  200: AUTH_STATUS_CODE.LOGGED_IN,
  401: AUTH_STATUS_CODE.NOT_LOGGED,
  403: AUTH_STATUS_CODE.NEED_PERMISSION,
};

/**
 * @param {Promise} promise
 * @return {Promise<any>}
 */
const withSuccess = promise => {
  return promise.then(res => {
    if (res && !res.success) throw new Error(res.message);

    return res;
  });
};

/**
 * @param {Function} handler
 * @return {number}
 */
const createTimeout = handler => setTimeout(handler, TIMEOUT_DEFAULT);

/**
 * @param {string} email
 * @return {Promise<any>}
 */
const getAuthChallenge = email => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth`, {
      email,
    }),
  );
};

/**
 * @param {string} password
 * @param {string} originHost
 * @return {Promise<any>}
 */
const setAuthPermission = async (password, originHost) => {
  const res = await request.post(`${identityBaseUrl}/auth/permission`, {
    password,
    originHost,
  });
  return res;
};

/**
 * @param {object} params
 * @param {string} params.email
 * @param {string} params.code
 * @param {string} params.password
 * @param {boolean} params.isSignUp
 * @param {boolean} params.isRemember
 * @return {Promise<any>}
 */
const authWithCode = async ({
  email,
  code,
  password,
  isSignUp,
  isRemember,
}) => {
  const url = isSignUp
    ? `${identityBaseUrl}/auth/signup`
    : `${identityBaseUrl}/auth/token`;

  return withSuccess(
    request.post(url, {
      email,
      code,
      password,
      remember: isRemember,
    }),
  );
};

/**
 * @param {string} idToken
 * @return {Promise<any>}
 */
const authWithGoogle = idToken =>
  request
    .post(`${identityBaseUrl}/auth/google`, {
      token: idToken,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

/**
 * @param {string} code
 * @return {Promise<any>}
 */
const authWithGitHub = code =>
  request
    .post(`${identityBaseUrl}/auth/github`, {
      code,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

const logout = () => request.post(`${identityBaseUrl}/logout`);

/**
 * @return {Promise<{expiresAt: number, status: string, hash: string}>}
 */
const getAuthStatus = async () => {
  try {
    const { expiresAt, hash } = await request.get(
      `${identityBaseUrl}/auth/check`,
    );
    return {
      status: AUTH_STATUS_CODE.LOGGED_IN,
      hash,
      expiresAt,
    };
  } catch (e) {
    const statusCode = get(e, ['response', 'status']);
    const status = CODE_TO_STATUS[statusCode] || AUTH_STATUS_CODE.NOT_LOGGED;

    const hash = get(e, ['response', 'data', 'hash'], '');
    const expiresAt = get(e, ['response', 'data', 'expiresAt'], 0);
    return {
      status,
      hash,
      expiresAt,
    };
  }
};

/**
 * @return {Promise<number>}
 */
const waitLogin = () =>
  new Promise((resolve, reject) => {
    /* eslint-disable-next-line */
    const handler = async function() {
      try {
        const { status } = await getAuthStatus();

        if (
          status === AUTH_STATUS_CODE.LOGGED_IN ||
          status === AUTH_STATUS_CODE.NEED_PERMISSION
        ) {
          return resolve();
        }

        createTimeout(handler);
      } catch (err) {
        return reject(err);
      }
    };

    return handler();
  });

/**
 * @param {string} email
 * @return {Promise<string>}
 */
const getRecoveryIdentifier = email =>
  request
    .get(`${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`)
    .then(res => {
      if (!res.success) throw new Error(res.message);

      return res.message;
    });

/**
 * @param {string} email
 * @return {Promise<number>}
 */
const sendEmailCode = async email => {
  const { timeout } = await request.post(`${identityBaseUrl}/auth/code`, {
    email,
  });

  return timeout;
};

/**
 * Send sms with code for disabling otp
 * @param {string} email
 * @returns {Promise<any>}
 */
const sendOtpRecoverSms = async email => {
  try {
    return withSuccess(
      request.get(
        `${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`,
      ),
    );
  } catch (error) {
    const { response = {} } = error;
    const { status } = response;
    error.code = status;

    throw error;
  }
};

/**
 * Disable otp setting
 * @param {object} param
 * @param {string} param.email
 * @param {string} param.code
 * @returns {Promise<any>}
 */
const disableOtpViaSms = async ({ email, code }) => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth/recover`, {
      email,
      code,
    }),
  );
};

/**
 * @return {string}
 */
const getSeedTemplateUrl = () => `${identityBaseUrl}/auth/recover/template`;

export default {
  getAuthStatus,
  setAuthPermission,
  getAuthChallenge,
  authWithCode,
  authWithGoogle,
  authWithGitHub,
  sendEmailCode,
  logout,
  waitLogin,
  getRecoveryIdentifier,
  sendOtpRecoverSms,
  disableOtpViaSms,
  getSeedTemplateUrl,
};
