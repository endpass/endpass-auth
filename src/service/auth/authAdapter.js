import { CHALLENGE_TYPES } from '@/constants';

function getChallengeType(challenge) {
  if (challenge === 'sms') {
    return CHALLENGE_TYPES.SMS_OTP;
  }

  if (challenge === 'otp') {
    return CHALLENGE_TYPES.APP_OTP;
  }

  return CHALLENGE_TYPES.EMAIL_OTP;
}

export default function(authData) {
  return {
    ...authData,
    challenge: {
      ...authData.challenge,
      challengeType: getChallengeType(authData.challenge),
    },
  };
}
