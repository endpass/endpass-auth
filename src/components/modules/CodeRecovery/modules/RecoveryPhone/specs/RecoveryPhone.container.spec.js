import { shallowMount, createLocalVue } from '@vue/test-utils';
import { code } from '@unitFixtures/auth';
import RecoveryPhoneContainer from '@/components/modules/CodeRecovery/modules/RecoveryPhone/RecoveryPhone.container';
import NoPhone from '@/components/modules/CodeRecovery/modules/RecoveryPhone/modules/NoPhone';
import WithPhone from '@/components/modules/CodeRecovery/modules/RecoveryPhone/modules/WithPhone';
import setupI18n from '@/locales/i18nSetup';
import LoadingScreen from '@/components/common/LoadingScreen';

const localVue = createLocalVue();
const i18n = setupI18n(localVue);

describe('RecoveryPhoneContainer', () => {
  let wrapper;

  const wrapperFactory = (options, props) => {
    return shallowMount(RecoveryPhoneContainer, {
      localVue,
      i18n,
      propsData: {
        error: '',
        isLoading: false,
        isPhoneExist: true,
        ...props,
      },
      sync: false,
      ...options,
    });
  };

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = wrapperFactory();
  });

  describe('render', () => {
    it('should correctly render component', () => {
      expect(wrapper.name()).toBe('RecoveryPhoneContainer');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render loading', () => {
      wrapper = wrapperFactory(null, {
        isLoading: true,
      });

      expect(wrapper.find(LoadingScreen).attributes().isloading).toBeTruthy();
    });

    it('should render no phone', () => {
      wrapper = wrapperFactory(null, {
        isPhoneExist: false,
      });

      expect(wrapper.find(NoPhone).exists()).toBe(true);
    });

    it('should render with phone', () => {
      wrapper = wrapperFactory(null, {
        isPhoneExist: true,
      });

      expect(wrapper.find(WithPhone).exists()).toBe(true);
    });
  });

  describe('behavior', () => {
    it('should emit complete event', () => {
      expect(wrapper.emitted().submit).toBeUndefined();

      wrapper.find(WithPhone).vm.$emit('submit', { code });

      expect(wrapper.emitted().submit.length).toBe(1);
      expect(wrapper.emitted().submit[0]).toEqual([{ code }]);
    });

    it('should emit send-code event', () => {
      expect(wrapper.emitted()['send-code']).toBeUndefined();

      wrapper.find(WithPhone).vm.$emit('send-code');

      expect(wrapper.emitted()['send-code'].length).toBe(1);
      expect(wrapper.emitted()['send-code'][0]).toEqual([]);
    });

    it('should emit recovery cancel event', () => {
      expect(wrapper.emitted().cancel).toBeUndefined();

      wrapper.find(WithPhone).vm.$emit('cancel');

      expect(wrapper.emitted()['recovery-cancel'].length).toBe(1);
      expect(wrapper.emitted()['recovery-cancel'][0]).toEqual([]);
    });
  });
});
