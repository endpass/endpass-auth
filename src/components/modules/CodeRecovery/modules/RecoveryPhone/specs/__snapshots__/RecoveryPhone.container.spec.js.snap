// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecoveryPhoneContainer render should correctly render component 1`] = `
<loading-screen-stub>
  <withphone-stub error=""></withphone-stub>
</loading-screen-stub>
`;
